R/nearest_stations_imgw.R

```diff
@@ -5,83 +5,92 @@
-  if (length(point)>2) {
+nearest_stations_imgw = function(type = "meteo",
+                                 rank = "synop",
+                                 year = 2018,
+                                 add_map = TRUE,
+                                 point = NULL,
+                                 no_of_stations = 50,
+                                 ...) {
+  if (length(point) > 2) {
     stop("Too many points for the distance calculations. Please provide just one pair of coordinates (e.g. point = c(17,53))")
-  } else if (length(point)<2 | length(point) == 0) {
-    message("
-            The point argument should have two coordinates.
+  } else if (length(point) < 2) {
+    message("The point argument should have two coordinates.
             We will provide nearest stations for mean location of all available stations.
-            To change it please change the `point` argument c(LON,LAT)" )
+            To change it please change the `point` argument c(LON,LAT)")
     Sys.sleep(2)
   }
   
-  
-  if (max(year)>=as.integer(substr(Sys.Date(),1,4))-1) {
+  if (point[1] > 180) {
+    stop("x should be longitude")
+  } else if (point[2] > 90) {
+    stop("y should be latitude")
+  }
+
+  if (max(year) >= as.integer(substr(Sys.Date(), 1, 4)) - 1) {
     message("Data cannot be provided for this repository. Please check the available records at: \n
             https://danepubliczne.imgw.pl/data/dane_pomiarowo_obserwacyjne/")
   }
   
-  if (type == "meteo"){
-    result = unique(meteo_imgw_monthly(rank = rank, year = year, coords = T)[,c(2:5)])
-  } else if (type=="hydro"){
-    result = unique(hydro_imgw_annual(year = year, coords = T)[,c(1:4)])
+  if (type == "meteo") {
+    result = unique(meteo_imgw_monthly(rank = rank, year = year, coords = TRUE)[, c(2:5)])
+  } else if (type == "hydro") {
+    result = unique(hydro_imgw_annual(year = year, coords = TRUE)[, c(1:4)])
   } else {
     stop("You've provided wrong type argument; please use: \"meteo\", or \"hydro\"")
   }
 
-  if (dim(result)[1]==0) {
+  if (nrow(result) == 0) {
     stop("Propobly there is no data in the downloaded object. Please check available records:  
         https://danepubliczne.imgw.pl/data/dane_pomiarowo_obserwacyjne/")
-  } 
-  if (is.null(point)){
+  }
+  
+  if (is.null(point)) {
   # workaround for different column names:
-    if(any(colnames(result)=="LON")) point = c(round(mean(result$LON,na.rm=T),2),round(mean(result$LAT,na.rm=T),2))
-    if(any(colnames(result)=="X")) point = c(round(mean(result$X,na.rm=T),2),round(mean(result$Y,na.rm=T),2))
+    if (any(colnames(result) == "LON"))
+      point = c(round(mean(result$LON, na.rm = TRUE), 2),
+                round(mean(result$LAT, na.rm = TRUE), 2))
+    if (any(colnames(result) == "X"))
+      point = c(round(mean(result$X, na.rm = TRUE), 2),
+                round(mean(result$Y, na.rm = TRUE), 2))
   }
   
+  dist_vec = double(nrow(result))
+  for (i in seq_along(dist_vec)) {
```

Comment:
please vectorise if possible (if not - leave as is)

Reply:
OK, I will vectorise this, but please note that vector preallocation and value assignment is faster and requires less memory than vectorization. On the other hand, there are probably not too many locations (thousands?), so it shouldn't be a problem.

Reply:
I agree, that can be left as is

Reply:
Just in case, reference for future:

```r
dist_vec = apply(X = result[, 2:3], MARGIN = 1, FUN = spheroid_dist, p1 = point)
```